### ISSUE TITLE
Preprocessing of dataset to obtain source.*, target.* and vocab.* files from original dataset.

### ISSUE THREAD

**Issue by sobalgi** [ysu1989/CrossSemparse#2]
Hello Mr. Su,

Thanks a lot for providing your original implementation. I really appreciate your quick response.
Initial run seems to be working fine (no code errors)!!!

I'll run the batch script with gpu and hopefully try to reproduce your original results.
Can you please let me know how long will it take for complete in-domain and cross-domain training and testing on gpu?

Also what does gold canonical utterance and top canonical utterence imply into testing logs?

Also I have gone through you code. I see that you have source.*, target.*, vocab* files already present for each domain. I was wondering if you have some additional preprocessing code used to create tgese files from the original overnight dataset files.

It would be very helpful to have this file as I'm trying to implement it myself.

Many thanks in advance.
Best regards,
Sourabh Balgi,
IISc, Bangalore.

_Originally posted by @sobalgi in https://github.com/ysu1989/CrossSemparse/issues/1#issuecomment-423472546_

**Comment by ysu1989** [comment]
Hi Sourabh,

On 1080 Ti the batch run script can be finished in about 10 hours.

In the testing log, "gold canonical utterance" is the ground truth, and the top canonical utterance is the top 1 canonical utterance predicted by the current model.

The vocab creation can be found in src/data_utils.py. The creation of the word embedding files are in a separate file that is not included. The logic of that function is basically taking a vocab file as input, looking up (or randomly initializing if missing) in pre-trained `word2vec` embedding, and standardizing/normalizing accordingly. The pre-trained `word2vec` embedding files are huge and we don't want to include in the repo. But with the `Gensim` library it should be fairly straightforward to implement this function. You can refer to line 267 in src/parser.py for details about how this embedding creation function is called when missing.

**Comment by sobalgi** [comment]
Hello Mr. Su,
Thank you for your response.

I have gone through the data_utils.py code.
I get the vocabulary creation, embedding standardization and normalization, token to index mapping of source and target files. 

But what I did not understand is how the source.train, source.valid, source.test, target.train, target.valid and target.test are obtained.
Because the original overnight dataset has just .groups and .examples files.

The (source/target).(train/valid/test).id files are obtained from data_utils.py. I get this part but not how (source/target).(train/valid/test) files with the input and corresponding canonical utterance from .groups file.

Many thanks,
Sourabh Balgi

**Comment by sobalgi** [comment]
Also I'm trying to reproduce your results, but so far with 2 executions I have not got the desired accuracies. In fact first execution exec0 had a lower cross-domain accuracy(78%) than in-domain accuracy(79%) !!!!

So I had to run the batch again. Second execution exec1 had a bit higher cross-domain accuracy(79%) than in-domain accuracy(77%).
Note : The accuracies mentioned above are with **word2vec + ES + X** and **word2vec + ES + I** methods respectively. Original result from the paper is 80% and 78% respectively.

I'm still trying to run more executions to average the accuracy and find the standard deviation of accuracy for a better generalized metric.

Best regards,
Sourabh Balgi

**Comment by ysu1989** [comment]
Hi Sourabh,

I just added the data processing part (scripts/process_data.py and scripts/process_data.sh) for creating the source.*, target.*, and candidates.json files. It added dependencies on jsonpickle and Stanford CoreNLP (for tokenization). The logic of this data processing is straightforward except the extracting of candidates, where we need to look at the execution result files (e.g., calendar.result) to find the denotation of logical forms. To use it, you need to set up Stanford CoreNLP on port 9000 of localhost (see src/nlp_service.py). Also note that the split of training and validation is random, so each run will produce a different split. 

Thanks for letting me know your results! Because of the small number of testing examples in Overnight (100-200 for each domain), there is some variance across different runs. I only ran the experiments in full once. Please let me know your results averaged over multiple runs (both before and after re-training)!

**Comment by sobalgi** [comment]
Awesome !!! This clarifies most of the things for me. Since the original code was not available earlier, I had written code myself to extract the input and it's corresponding canonical utterance from .groups file of overnight dataset. I had used the train and test utterances from .examples file of the original dataset. One thing I could not get matching was your size of training and testing instances. Hence I requested for your code. 

Now that I know training and validation splits are random, this behavior is understandable. 

Because of shuffling of instances in training, each model is unique and hence the variance in accuracy is also justifiable. I have already started several executions and got the accuracies for most. I'll let you know the details once I have the complete accuracies from all the runs. 

Thanks again for you quick response. I appreciate it a lot.

Best regards,
Sourabh Balgi
